File: vm/Sources/SafeYoloVM/BridgeFilter.swift
import Foundation

/// Enable pf IP filtering on a macOS bridge interface.
///
/// Apple's bridge interfaces have IP filtering disabled by default,
/// which means pf rules on the bridge are ignored. This uses the
/// BRDGSFILT ioctl to enable the IFBF_FILT_USEIPF flag, allowing
/// pf to intercept packets traversing the bridge.
///
/// Requires root privileges (run via sudo).
enum BridgeFilter {

    // From bsd/net/if_bridgevar.h in XNU kernel source
    private static let BRDGSFILT: UInt = 24
    private static let IFBF_FILT_USEIPF: UInt32 = 0x00000001

    // From sys/ioctl.h — SIOCSDRVSPEC
    // SIOCSDRVSPEC = _IOW('i', 124, struct ifdrv)
    // = IOC_IN(0x80000000) | (sizeof(ifdrv)=40=0x28 << 16) | ('i'=0x69 << 8) | 124=0x7C
    // = 0x8028697C
    private static let SIOCSDRVSPEC: UInt = 0x8028697C

    /// Layout must match struct ifdrv from <net/if_var.h>
    /// struct ifdrv {
    ///     char    ifd_name[IFNAMSIZ];  // 16 bytes
    ///     unsigned long ifd_cmd;        // 8 bytes (64-bit)
    ///     size_t  ifd_len;              // 8 bytes
    ///     void   *ifd_data;             // 8 bytes
    /// };
    struct ifdrv {
        var ifd_name: (CChar, CChar, CChar, CChar, CChar, CChar, CChar, CChar,
                       CChar, CChar, CChar, CChar, CChar, CChar, CChar, CChar) = (0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0)
        var ifd_cmd: UInt = 0
        var ifd_len: UInt = 0
        var ifd_data: UnsafeMutableRawPointer? = nil
    }

    struct ifbrparam {
        var ifbrp_filter: UInt32 = 0
    }

    static func enableIPFilter(on bridge: String) -> Bool {
        let sock = socket(AF_INET, SOCK_DGRAM, 0)
        guard sock >= 0 else {
            fputs("Error: cannot create socket: \(String(cString: strerror(errno)))\n", stderr)
            return false
        }
        defer { close(sock) }

        var param = ifbrparam(ifbrp_filter: IFBF_FILT_USEIPF)
        var drv = ifdrv()
        drv.ifd_cmd = BRDGSFILT
        drv.ifd_len = UInt(MemoryLayout<ifbrparam>.size)

        // Copy bridge name into ifd_name tuple
        withUnsafeMutablePointer(to: &drv.ifd_name) { namePtr in
            let raw = UnsafeMutableRawPointer(namePtr).assumingMemoryBound(to: CChar.self)
            bridge.utf8CString.withUnsafeBufferPointer { buf in
                let len = min(buf.count, 16)
                for i in 0..<len {
                    raw[i] = buf[i]
                }
            }
        }

        // Point ifd_data at our param struct
        let result = withUnsafeMutablePointer(to: &param) { paramPtr -> Int32 in
            drv.ifd_data = UnsafeMutableRawPointer(paramPtr)
            return withUnsafeMutablePointer(to: &drv) { drvPtr -> Int32 in
                ioctl(sock, UInt(SIOCSDRVSPEC), drvPtr)
            }
        }

        if result < 0 {
            fputs("Error: ioctl BRDGSFILT on \(bridge): \(String(cString: strerror(errno)))\n", stderr)
            return false
        }

        fputs("ipfilter enabled on \(bridge)\n", stderr)
        return true
    }
}

File: vm/bridge-filter.c
/*
 * bridge-filter: Enable pf IP filtering on a macOS bridge interface.
 *
 * macOS bridge interfaces have ipfilter disabled by default, meaning pf
 * rules on the bridge are silently ignored. This tool uses the BRDGSFILT
 * ioctl to set the IFBF_FILT_USEIPF flag from XNU's if_bridgevar.h.
 *
 * Usage: sudo bridge-filter <bridge-name>
 *
 * Build: cc -o bridge-filter bridge-filter.c
 */

#include <sys/ioctl.h>
#include <sys/socket.h>
#include <net/if.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>

/* From bsd/net/if_bridgevar.h (XNU kernel source) */
#define BRDGSFILT   24
#define BRDGGFILT   23

#define IFBF_FILT_USEIPF    0x00000001  /* run pf hooks on bridge */
#define IFBF_FILT_MEMBER    0x00000002  /* run pf hooks on members */
#define IFBF_FILT_ONLYIP    0x00000004  /* only filter IP packets */

/*
 * SIOCSDRVSPEC = _IOW('i', 124, struct ifdrv) = 0x8028697C
 * SIOCGDRVSPEC = _IOWR('i', 123, struct ifdrv) = 0xC0286977
 */
#ifndef SIOCSDRVSPEC
#define SIOCSDRVSPEC    0x8028697C
#endif
#ifndef SIOCGDRVSPEC
#define SIOCGDRVSPEC    0xC0286977
#endif

struct ifdrv {
    char        ifd_name[IFNAMSIZ];
    unsigned long   ifd_cmd;
    size_t      ifd_len;
    void        *ifd_data;
};

struct ifbrparam {
    union {
        uint32_t ifbrpu_int32;
        uint16_t ifbrpu_int16;
        uint8_t  ifbrpu_int8;
    } ifbrp_ifbrpu;
};
#define ifbrp_filter    ifbrp_ifbrpu.ifbrpu_int32

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <bridge-name>\n", argv[0]);
        return 1;
    }

    const char *bridge = argv[1];
    int s = socket(AF_INET, SOCK_DGRAM, 0);
    if (s < 0) {
        perror("socket");
        return 1;
    }

    /* First, read current filter flags */
    struct ifbrparam gp;
    memset(&gp, 0, sizeof(gp));
    struct ifdrv gdrv;
    memset(&gdrv, 0, sizeof(gdrv));
    strlcpy(gdrv.ifd_name, bridge, IFNAMSIZ);
    gdrv.ifd_cmd = BRDGGFILT;
    gdrv.ifd_len = sizeof(gp);
    gdrv.ifd_data = &gp;

    if (ioctl(s, SIOCGDRVSPEC, &gdrv) < 0) {
        fprintf(stderr, "ioctl BRDGGFILT on %s: %s\n", bridge, strerror(errno));
        close(s);
        return 1;
    }

    uint32_t current = gp.ifbrp_filter;
    if (current & IFBF_FILT_USEIPF) {
        fprintf(stderr, "ipfilter already enabled on %s (flags=0x%x)\n", bridge, current);
        close(s);
        return 0;
    }

    /* Enable USEIPF flag */
    struct ifbrparam sp;
    memset(&sp, 0, sizeof(sp));
    sp.ifbrp_filter = current | IFBF_FILT_USEIPF;

    struct ifdrv sdrv;
    memset(&sdrv, 0, sizeof(sdrv));
    strlcpy(sdrv.ifd_name, bridge, IFNAMSIZ);
    sdrv.ifd_cmd = BRDGSFILT;
    sdrv.ifd_len = sizeof(sp);
    sdrv.ifd_data = &sp;

    if (ioctl(s, SIOCSDRVSPEC, &sdrv) < 0) {
        fprintf(stderr, "ioctl BRDGSFILT on %s: %s\n", bridge, strerror(errno));
        close(s);
        return 1;
    }

    fprintf(stderr, "ipfilter enabled on %s (flags=0x%x → 0x%x)\n",
            bridge, current, sp.ifbrp_filter);
    close(s);
    return 0;
}
